# Data Cleaning with Python Pandas

## Scenario

The operations manager at a mid-sized **customer engagement firm** specializing in telemarketing and customer support services has identified significant inefficiencies in their call operations. The firm relies heavily on accurate and up-to-date customer data to enable smooth and productive client interactions. However, the current **customer call list** is plagued by data quality issues that negatively impact call agents' productivity and operational performance.

These issues include:
- Duplicate entries, leading to redundant calls.
* Missing values (e.g., phone numbers or customer names), making some entries unusable.
+ Inconsistent formatting (e.g., phone numbers stored in multiple formats, incorrect casing in names).

Addressing these challenges is critical to ensuring operational efficiency and enhancing the firm’s ability to provide seamless client support.

## Stakeholders

1. **Operations Manager:** Concerned about improving the overall efficiency of call operations and ensuring accurate, high-quality customer data.
2. **Call Agents:** The primary users of the customer call list who rely on clean and complete data for their daily workflows.
3. **IT/Data Team:** Responsible for managing, maintaining, and updating the customer database.
4. **Executive Team:** Interested in achieving improved productivity, reduced call redundancies, and higher customer satisfaction.

## Purpose

The primary goal of this project is to provide the client with a clean, standardized, and actionable customer call list that eliminates redundancies, corrects inconsistencies, and ensures data readiness for immediate use by call agents.

### Objectives:

1. Identify and remove duplicate records to avoid redundancy.
2. Handle missing values to ensure all essential fields required for making calls are complete or flagged appropriately.
3. Standardize formatting for consistency across fields (e.g., phone numbers, names, and addresses).
4. Remove unnecessary columns and rows that do not contribute to the call agent’s workflow.

## Data Sources

The primary data source for this project is the **customer call list.xlsx**, which contains the following fields:

- CustomerID
- First Name
- Last Name
- Phone Number
- Address
- Paying Customer Status
- Do_Not_Contact Status
- Not_Useful_Column

## Deliverables

1. **Cleaned Dataset:** **Excel spreadsheet (.xlsx)** with standardized and deduplicated customer call list. Only actionable fields to be retained.
2. **Data Cleaning Report:** **PDF document** containing summary of the cleaning process

## Business Metrics

To measure whether the **goal of delivering a clean, standardized, and actionable customer call list** is achieved, the firm can track the following **business metrics:**

### 1. Reduction in Duplicate Calls

**Metric:** Percentage decrease in duplicate calls made by call agents.

- **Why it matters:** Eliminating duplicate records ensures that call agents are not wasting time contacting the same customer multiple times.
- **How to measure:** Compare the number of duplicate calls before and after implementing the cleaned call list.

### 2. Increase in Call Connection Rate
**Metric:** Percentage increase in successful call connections.

- **Why it matters:** Standardizing phone numbers and removing invalid entries will improve the likelihood of calls reaching valid customers.
- **How to measure:** Track the ratio of successful connections (valid calls) to total calls before and after the project.

### 3. Reduction in Invalid or Incomplete Records
**Metric:** Percentage decrease in unusable records (e.g., missing phone numbers or names).

- **Why it matters:** Handling missing values ensures that call agents have complete and actionable data to work with.
- **How to measure:** Compare the number of records with missing values before and after the data cleaning process.

### 4. Agent Productivity
**Metric:** Number of calls made per agent per hour/day.

- **Why it matters:** A clean and streamlined dataset allows agents to spend less time dealing with incomplete or invalid records, enabling them to make more calls.
- **How to measure:** Track the average number of calls completed per agent over a specific time period.

### 5. Reduction in Call Handling Time
**Metric:** Average time spent per call (reduction in handling delays caused by data issues).

- **Why it matters:** With standardized and accurate data, call agents can quickly identify and call the right customers, reducing the time wasted searching or verifying information.
- **How to measure:** Compare average call handling time before and after the dataset is cleaned.

### 6. Customer Response Rate
**Metric:** Percentage increase in responses from customers (e.g., call completions or callbacks).

- **Why it matters:** A clean call list with valid data ensures agents reach real customers, improving customer engagement and response rates.
- **How to measure:** Compare the number of customer responses to total calls made before and after implementing the cleaned list.

### Summary Table of Metrics

| Metric	| Goal	 | Measurement |
| :-------- | :----- | :---------- |
| Reduction in Duplicate Calls	| 100% removal of duplicate calls |	Compare pre/post duplicate call records. |
| Increase in Call Connection Rate | Improve by X%	| Measure ratio of successful calls. |
| Reduction in Invalid Records | Reduce unusable records by X% | Count records with missing values. |
| Agent Productivity | Increase calls per hour/day	| Average calls per agent before/after. |
| Reduction in Call Handling Time |	Decrease handling delays | Measure average call time. |
| Customer Response Rate | Increase engagement by X% | Track successful responses per call. |

Once the cleaned call list is implemented, these metrics can be tracked over a defined time period (e.g., weekly or monthly) to evaluate improvements. This will help the firm **quantify the success of the project** and demonstrate its impact on operational efficiency and agent performance.


## Installing and Loading Packages

In [1]:
# To update a package, run the following command in the terminal or command prompt:
# pip install -U package_name

# To install an exact version of a package, run the following command in the terminal or command prompt:
# !pip install package_name==desired_version

# After installing or updating the package, restart the Jupyter notebook.

#1. Install the `watermark` package.
#1.a. This package is used to record the versions of other packages used in this Jupyter notebook.
!pip install -q -U watermark

In [2]:
import os
import pandas as pd

# Suppress warnings
import warnings
warnings.filterwarnings('ignore')

In [3]:
%reload_ext watermark
%watermark -a "AmeduStephen"

Author: AmeduStephen



## Data Acquisition

In [10]:
# Get the directory of the script's location, assumed here to be '../notebooks' and to be on the same folder level with '../data'
script_dir = os.getcwd()
# Please note that os.getcwd() depends on the current working directory, which might not always align with the script's location  

# Navigate to the parent folder
parent_dir = os.path.abspath(os.path.join(script_dir, ".."))

# Construct the path to the Excel file in the desired relative location
raw_data_path = os.path.join(parent_dir, "data", "raw", "Customer Call List.xlsx")

# Read the Excel file into a DataFrame
df = pd.read_excel(raw_data_path)

## Data Profiling

In this section, we will try to find information about the data types used, look at the datasets for inconsistent data formats, and identify missing values and duplications

In [13]:
# Check data type to get the general overview of the data
print("\nData Information:")
print(df.info())


Data Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   CustomerID         21 non-null     int64 
 1   First_Name         21 non-null     object
 2   Last_Name          20 non-null     object
 3   Phone_Number       19 non-null     object
 4   Address            21 non-null     object
 5   Paying Customer    21 non-null     object
 6   Do_Not_Contact     17 non-null     object
 7   Not_Useful_Column  21 non-null     bool  
dtypes: bool(1), int64(1), object(6)
memory usage: 1.3+ KB
None


In [26]:
# check number of rows and columns
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 21
Number of columns: 8


In [27]:
# Display first x (e.g. 10) rows
#df.head(10)

In [28]:
# Display last 10 rows
#df.tail(10)

In [29]:
# Fetch random samples from dataset (50%)
#df_sample = df.sample(frac=0.50, random_state=100)
#df_sample

In [30]:
# Display the entire dataset for visual inspection, since it is not much any way. Otherwise, we would use df.head() and df.tail()
df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   
20        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes            NaN   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   
20            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True  
20               True

In [32]:
# check for duplicate rows
print("Number of duplicate rows:")
print("-"*30)
df.duplicated().sum()

Number of duplicate rows:
------------------------------


1

In [34]:
# Show duplicates including the first occurrence
duplicates_incl_first = df[df.duplicated(keep=False)]
print("Duplicates Including First Occurrence:\n", duplicates_incl_first)

Duplicates Including First Occurrence:
     CustomerID First_Name  Last_Name  Phone_Number  \
19        1020     Anakin  Skywalker  876|678|3469   
20        1020     Anakin  Skywalker  876|678|3469   

                        Address Paying Customer Do_Not_Contact  \
19  910 Tatooine Road, Tatooine             Yes              N   
20  910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
19               True  
20               True  


In [35]:
# determine if dataset has null values 

df.isnull().sum()

CustomerID           0
First_Name           0
Last_Name            1
Phone_Number         2
Address              0
Paying Customer      0
Do_Not_Contact       4
Not_Useful_Column    0
dtype: int64

## Data understanding and cleaning strategy

Here is our understanding of the data dictionary:

| Variable Name	    |  Data Type	  | Description	                                             |  Example Value        |
| :---------------- | :-------------- | :------------------------------------------------------- | :-------------------- |
| CustomerID	    | Integer	      | Unique identifier for each customer.	                 | 1001                  |
| First_Name        | String	      | First name of the customer.	                             | Frodo                 |
| Last_Name	        | String	      | Last name or family name of the customer.	             | Baggins               |
| Phone_Number	    | String	      | Customer's phone number (includes inconsistent formats). | 123-545-5421          |
| Address	        | String          | Customer's physical address.	                         | 123 Shire Lane, Shire |
| Paying Customer   | String (Yes/No) | Indicates if the customer is a paying customer.	         | Yes                   |
| Do_Not_Contact    | String (Yes/No) | Indicates if the customer has opted out of contact.	     | No                    |
| Not_Useful_Column | Boolean	      | Column with unclear relevance or no apparent use.	     | TRUE                  |

From the data profiling stage, we can see that:

1. One row is duplicated; we need to remove that.
2. The column **'First_Name'** has one null ('NaN' - Not a Number) value, and apparent errors like '/', '..' preceding some names; we need to clean it up.
3. The column **'Phone_Number'** has null ('NaN') values and variations like 'N/a', along with inconsistent formats like 123-545-5421, 123/643/9775, and 876|678|3469; we need to clean and standardize it.
4. The column **'Do_Not_Contact'** has four null ('NaN') values; we need to clean it up.
5. Both columns **'Do_Not_Contact'** and **'Paying Customer'** mix 'Yes/No' values with variations like 'Y' or 'N'; we need to standardize them.
6. The column **'Not_Useful_Column'**, as the name suggests, seems irrelevant and not fit for purpose; we need to drop it

We propose to split the **'Address'** column into **'Street_Address'**, **'State'** and **'Zip_Code'** for clarity, and to replace all 'NaN' or 'N/a' values with ''. Finally, we will drop all rows with **'Do_Not_Contact' is equal to 'Y'** or **'Phone_Number' equal to ''**.

## Data Cleansing

## Export Cleaned Dataset

In [ ]:
# saving the data file as csv to current directory location
#path_to_file  = abs_path + "/data/cleaned/cleaned_customer_call_list.xlsx"
#df.to_excel(path_to_file,index=False)

# The End